### ISSUE TITLE
cc65: compiler's if condition judgment error

### ISSUE THREAD

**Issue by gb-2312** [cc65/cc65#1259]
In the process of using cc65 to develop, I found an `if` logic issue about the cc65 compiler, such as this:
```c
#include "xxx.h"
//...

// unsigned-char!
unsigned char vel_y;

//...
void main(void) 
{
	vel_y = 0;

	// if (95 > (99 - 0)) really???
	if (95 > (99 - vel_y)) {
		put_str(NTADR_A(2,2),"11111");
	}

	else {
		put_str(NTADR_A(2,2),"22222");
	}
}
```
The execution error-result of the logic is `11111`, but the correct logic should be `22222`.

compiler: `nes.lib`
cc65-hash(version): b8ae5c28fe18ba5d5e0791eecded61d0a16f0d0e
OS: macos-10.15.6 /windows-server-2003
issue-project(**.bash and .bat files have already included**): https://github.com/gb-2312/cc65_issue

![Screen Shot 2020-09-17 at 10 13 09 PM](https://user-images.githubusercontent.com/435219/93485632-d9bf5780-f935-11ea-8461-b0cf5df2ac0f.png)

![Screen Shot 2020-09-17 at 10 20 34 PM](https://user-images.githubusercontent.com/435219/93485660-e17efc00-f935-11ea-88cf-ccfc8f2e1afc.png)

----

**_Update@Fri Sep 18 10:45:19 CST 2020:_** 
I have already located the issue version,

**last condition-correct version(hash)**: 57117fa687de59234d299d76f406f4e666844f1f
**first condition-error version(hash):**  7553b60ef07a53c797974a73627cd6bebcc8e8fe
**compiler: `nes.lib`(NES-console)**

and I using diff command to find [.s]differ:

```diff
angel@localhost cc65_issue % git diff example1.s 
diff --git a/example1.s b/example1.s
index ce70aea..57b0d3d 100644
--- a/example1.s
+++ b/example1.s
@@ -1,7 +1,7 @@
 ;
-; File generated by cc65 v 2.18 - Git 57117fa6
+; File generated by cc65 v 2.18 - Git 7553b60e
 ;
-       .fopt           compiler,"cc65 v 2.18 - Git 57117fa6"
+       .fopt           compiler,"cc65 v 2.18 - Git 7553b60e"
        .setcpu         "6502"
        .smart          on
        .autoimport     on
@@ -121,13 +121,15 @@ L0005:    jmp     incsp4
 ;
 ; if (95 > (99 - vel_y)) {
 ;
-       lda     #$5F
-       jsr     pusha0
        lda     #$63
        sec
        sbc     _vel_y
-       jsr     tosicmp0
-       bmi     L0015
+       sta     ptr1+1
+       lda     #$5F
+       cmp     ptr1+1
+       beq     L0030
+       lda     #$01
+L0030: bmi     L0015
        beq     L0015
 ;
 ; put_str(NTADR_A(2,2),"11111");
```

**Comment by polluks** [comment]
Hi!
I was trying sim65, it works as expected.
cc65 V2.18 - Git 74144c62

2020-09-17 16:43 GMT+02:00, Angel <notifications@github.com>:

**Comment by acqn** [comment]
Doe it work right with PR #1258?

**Comment by greg-king5** [comment]
> Does it work right with PR #1258?

Yep!

@gb-2312, cc65 now obeys the C standard's "int promotion rule".  Therefore, your code is converted into a `signed int` comparison test.  The current optimizer squeezes that into a `signed char` test, but the generated code is broken.  PR #1258 stops the squeeze.  The code stays as a `signed int` test.

If you want the behavior that you expect for an 8-bit controller, then you must force it explicitly!  You must cast both sides of the comparison to `unsigned char`:
```c
if ((unsigned char)95 > (unsigned char)(99 - vel_y)) {
```

**Comment by mrdudz** [comment]
hu? obviously that test must work without the casts when vel_y is 0

**Comment by gb-2312** [comment]
> Does it work right with PR #1258?

Hi, I've `checkout`, `make clean`, `make -j4`, and try using version(acqn->cc65->[b]StateTracking) of #1258 , 
**this works correct** :)

![Screen Shot 2020-09-18 at 11 40 15 AM](https://user-images.githubusercontent.com/435219/93552697-c945c600-f9a3-11ea-9fa9-42444d77d4f4.png)

.s(assembler code) screenshot(with highlight):

![Screen Shot 2020-09-18 at 11 43 14 AM](https://user-images.githubusercontent.com/435219/93552840-2b9ec680-f9a4-11ea-939e-a110994a90df.png)

**Comment by gb-2312** [comment]
> > Does it work right with PR #1258?
> 
> Yep!
> 
> @gb-2312, cc65 now obeys the C standard's "int promotion rule". Therefore, your code is converted into a `signed int` comparison test. The current optimizer squeezes that into a `signed char` test, but the generated code is broken. PR #1258 stops the squeeze. The code stays as a `signed int` test.
> 
> If you want the behavior that you expect for an 8-bit controller chip, then you must force it explicitly! You must cast both sides of the comparison to `unsigned char`:
> 
> ```c
> if ((unsigned char)95 > (unsigned char)(99 - vel_y)) {
> ```

Thank you very much, I understand! `integer promotions in C` is indeed a trap that is easy to make mistakes! but my project have a lot of similar code, wow!! LOL

**Comment by jmr** [comment]
> `if (95 > (99 - vel_y)) {`

I'm with the "hu?" crowd. 95 is an `int`. 99 is an `int`. `vel_y` is `unsigned char`; it is implicitly converted to `int`. `95 < 99`. I don't see what this has to do with the "usual arithmetic conversions". I also made several variations of self-contained test cases and they all worked.

I also don't see how the `tosicmp0` and `tosicmp` versions should give different results, so I must be missing something there. BTW, it would be much more convenient if you pasted the diff into the bug rather than using links to screenshots.

**Comment by gb-2312** [comment]
> > `if (95 > (99 - vel_y)) {`
> 
> I'm with the "hu?" crowd. 95 is an `int`. 99 is an `int`. `vel_y` is `unsigned char`; it is implicitly converted to `int`. `95 < 99`. I don't see what this has to do with the "usual arithmetic conversions". I also made several variations of self-contained test cases and they all worked.
> 
> I also don't see how the `tosicmp0` and `tosicmp` versions should give different results, so I must be missing something there. BTW, it would be much more convenient if you pasted the diff into the bug rather than using links to screenshots.

Thank you for your prompt, I have replaced the screenshot with the code.😄

**Comment by acqn** [comment]
It seems to me that the sequence
```asm
pha
txa
sbc     #$00
tax
pla
```
could be optimized into
```asm
bcs L0
dex
L0:
```
if the X result is used but the C/V/N/Z flags are unused. The C flag condition could be omitted if X was zero.

**Comment by acqn** [comment]
> 
> 
> > `if (95 > (99 - vel_y)) {`
> 
> I'm with the "hu?" crowd. 95 is an `int`. 99 is an `int`. `vel_y` is `unsigned char`; it is implicitly converted to `int`. `95 < 99`. I don't see what this has to do with the "usual arithmetic conversions". I also made several variations of self-contained test cases and they all worked.

The error was caused by an old bug in `OptPrecalc` that was fixed in PR #1258. The previous change in `OptStackOps` just generated a code pattern that exposed the bug.

> 
> I also don't see how the `tosicmp0` and `tosicmp` versions should give different results, so I must be missing something there. BTW, it would be much more convenient if you pasted the diff into the bug rather than using links to screenshots.

Those two versions (57117fa6 "the last good" and 0054443f "the fixed") both gave the expected output. The wrong version is 7553b60 the one in the OP that uses neither `tosicmp0` nor `tosicmp`.

**Comment by gb-2312** [comment]
Thank you everyone, now this problem has been solved, I have pull-rebase, recompile tested the latest version, there is no problem, now I closed this issue😊
